src/Namotion.Reflection/EnumerableExtensions.cs

```diff
@@ -41,9 +41,37 @@ public static IEnumerable<T> GetAssignableToTypeName<T>(this IEnumerable<T> obje
         /// <returns>May return null (not found).</returns>
         public static T? FirstAssignableToTypeNameOrDefault<T>(this IEnumerable<T>? objects, string typeName, TypeNameStyle typeNameStyle = TypeNameStyle.FullName)
         {
-            return objects != null ?
-                objects.GetAssignableToTypeName(typeName, typeNameStyle).FirstOrDefault() :
-                default;
+            if (objects != null)
+            {
+                foreach (var o in objects)
+                {
+                    if (o.GetType().IsAssignableToTypeName(typeName, typeNameStyle))
+                    {
+                        return o;
+                    }
+                }
+            }
+            return default;
+        }
+
+        /// <summary>Tries to get the first object which is assignable to the given type name.</summary>
+        /// <param name="objects">The objects.</param>
+        /// <param name="typeName">Type of the attribute.</param>
+        /// <param name="typeNameStyle">The type name style.</param>
+        /// <returns>May return null (not found).</returns>
+        public static T? FirstAssignableToTypeNameOrDefault<T>(this T[]? objects, string typeName, TypeNameStyle typeNameStyle = TypeNameStyle.FullName)
```

Comment:
Is this actually faster? Doesnt it still use an enumerator, i.e. it's exactly the same?

Reply:
Foreach on arrays is fastest possible, arrays don't use enumerators unless they are accessed through IEnumerable. IList<> etc are bad because List<> would return struct Enumerator, but interface forces to box struct to IEnumerable. 

[There's huge difference with code gen](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACBmABJgJlwGFcBvAWAChdbdq7cBLAOzgG0BdXAfUQCd+iAJ64AvLhYBTAO64u5XAEYYuAqrwBfANwM6mJVAB0B7AB5WcAHy8pLAK6opQ4ABsp4ybPncyy1eq4Wro0dHq0OMxsuACCgiIxAMaJUsjIABQAlOThjFFwuIkA9vbREgAMIXl0AGZF/FKIiQAWuOkAbgLMcFKoUbwCQsKZuYyUodWMxaUFANQSTD2oVZOao/oA7IUlbCu0axP0h5GW+ErYSSlpWTmHjKfTZbiV67R1DU2tHV2Lvf08dkczkQbikIzudHGkzyjzmCyWezyB2hmC2sMRB2omiAA)

Reply:
Aha, it uses an "optimized" foreach...

![image](https://user-images.githubusercontent.com/2603405/145380595-2eddcbbb-6f0e-4963-a409-cb508f8d15f5.png)

Reply:
To avoid a "messy" API (ie. for a consumer unnecessary overloads), would it be an option to use this: 

```
if (objects is T[] arr) 
{
    foreach (var o in arr)
    {
        if (o.GetType().IsAssignableToTypeName(typeName, typeNameStyle))
        {
            return o;
        }
    }
}
else
{
    if (objects != null)
    {
        foreach (var o in objects)
        {
            if (o.GetType().IsAssignableToTypeName(typeName, typeNameStyle))
            {
                return o;
            }
        }
    }
    return default;
}
```

what speaks against that? (it's just faster transparently)

Reply:
I'll do that, good idea, that's actually something that some libs use to test for specific concrete type already to guide CLR for faster operation

Reply:
Addressed feedback in latest commit